### ISSUE TITLE
How to handle MutationRuleFunc for Update/Delete Ops?

### ISSUE THREAD

**Issue by adayNU** [facebook/ent#1179]
Using the example in the documentation for reference here.

Let's assume every `Schema` has the `TenantMixin` (meaning every `Schema` has an edge to `Tenant`).

It seems fairly trivial to handle a Mutation when the Op is Create (as done in the example). Here's a more generic version of the `DenyMismatchedTenants` function in the docs:

```go
// DenyMismatchedTenants checks that the mutation is only mutating data from the same tenent 
func DenyMismatchedTenants() privacy.MutationRule {
    // Create a rule, and limit it to create operations below.
    type TenantMutator interface {
        TenantID() (int, bool)
	TenantCleared() bool
        ID() (id int, exists bool)
        Client() *ent.Client
    }
    return privacy.MutationRuleFunc(func(ctx context.Context, m ent.Mutation) error {
        mut, ok := m.(TenantMutator)
        if !ok {
            return privacy.Denyf("unexpected mutation type %T", mut)
        }
        
        // Assume the user's tenant has been attached to the context and is retrievable via this function.
        var tid = tenant.FromContext(ctx) 
        // This first check isn't explicitly needed for a Create Op as the creation itself will return a Constraint error.
        // It is helpful for an Update Op where this is explicitly set, however, on Update it's also totally possible to
        // mutate anything simply by updating it's Tenant edge to tid (assuming no other logic checks that below).
        if id, ok := mut.TenantID(); !ok {
		return privacy.Denyf("no tenant ID edge set with mutation")
	} else if id != cust.ID {
		return privacy.Denyf("tenant ID mismatch")
	}

        // Again, this would fail a constraint check anyway, but here it is for posterity.
        if mut.TenantCleared() {
                return privacy.Denyf("no tenant ID edge set with mutation")
        } 

         /*
        From here - I'm at a loss on how to verify if an Update / Delete is valid.
        For example I've tried...
       
        id, ok := mutID()

       ...thinking I could use mut.Client() to find the thing being mutated w/ its Tenant edge.
       However in every case id has not been set.
        */

        // Skip to the next privacy rule (equivalent to return nil).
        return privacy.Skip
    }
}
```

Curious if it's just not possible today or if I'm just totally whiffing on this implementaiton.

**Comment by a8m** [comment]
Hey @adayNU, sorry for the delay - Sabbath mode 😃 

I hope you saw that in the `DenyMismatchedTenants` example, the rule is limited only for create operations:

```go
func DenyMismatchedTenants() privacy.MutationRule {
    rule := privacy.GroupMutationRuleFunc(func(ctx context.Context, m *ent.GroupMutation) error {
        // ...
    })
    // Evaluate the mutation rule only on group creation.
    return privacy.OnMutationOperation(rule, ent.OpCreate)
}
```

I guess I can improve this example by rejecting changes to the tenant edge.

Regarding your rule - let me know if I got it properly, you allow mutating the tenant edge, but you want to verify that it matches the tenant-id in user-context?

**Comment by adayNU** [comment]
Hey @a8m no problem! Wasn't expecting a response until Monday anyway :)

I'm realizing my description of the problem was pretty hard to follow. so I figured what better than do so with code. Lets say the example has a new `Schema` called `Dataset` which can have records added by users of any Tenant. When it comes to mutations, users should only be able to:
-Create a record belonging to their tenant
-Update a record that belongs to their tenant (and the mutation itself should not change/remove the tenant)
-Delete a record that belongs to their tenant

The goal is to write a `MutationRule` to accomplish the above. Here is my stab at what this specific example might look like [here](https://github.com/facebook/ent/compare/master...adayNU:example/privacy-tenant). Note: [here](https://github.com/facebook/ent/compare/master...adayNU:example/privacy-tenant#diff-836cbe34c3ff22f1328368b7eb11a369ba0f3d2afec60fcaee1015e9b36eb83dR111-R115) is the specific place where I've been running into issues -- it appears that for update/delete the ID field is not set.

Next, let's create a generic `MutationRule` to apply these rules to any `Schema` which has the `TenantMixin`. Here, would be an [example of this](https://github.com/facebook/ent/compare/master...adayNU:example/privacy-tenant-generic). Again this runs into the same potential issue.

I hope this better illustrates the issue i've run into (and the problem I'm trying to solve). Let me know if there is any more clarity I can provide.

**Comment by adayNU** [comment]
I also wrote a test to show the behavior I'm seeing since I figured that'd be helpful: https://github.com/adayNU/ent/commit/e849bd016f78f3dbd62352d496c5d85f5bb15c18

**Comment by adayNU** [comment]
Ok after digging into the code some more, it seems like one possible way to solve this is only support `UpdateOne` / `DeleteOne` Ops, and reject and `Delete`/`Update` Ops as the former set the id field in the mutation.

e.g. from my example code have a case the case statement instead look like:

```go
switch m.Op() {
case op.Create:
// Same as in my example.
case op.UpdateOne, op.DeleteOne:
// Same logic in my example (but for different Ops)
case op.Update, op.Delete:
    privacy.Denyf("unsupported mutation operation")
}
```

Curious if this is the only (currently possible) approach or if there is a better one.

**Comment by adayNU** [comment]
Also if that is the best solution, curious if you'd be open to implementing (I'd be happy to!) a way to more trivially get the `Tenant` edge from the mutation itself so that the more general example I added would be much less verbose (i.e. I don't need to switch on `m.Type()` and instead can add a method to the `TenantMutator` interface to get it.

Something in `ent/mutation.go` like:

```go
func (m *DatasetMutation) TenantQuery() (*TenantQuery, error) {
	if m.id == nil {
		return nil, errors.New("no id in mutation")
	}

	return m.Client().Dataset.Query().Where(dataset.ID(m.id)).QueryTenant()
}
```

**Comment by adayNU** [comment]
@a8m any thoughts here? My TL;DR is that it seems like `UpdateOne`/`DeleteOne` is the only way to do this today (which is fine) and wondering if it seems reasonable to add methods to mutations to more easily get an edge(s) in the `MutationRuleFunc`?

**Comment by a8m** [comment]
Hey @adayNU and sorry again for the delay - I'm a little bit busy lately with other parts (related to ent). 

Regarding your use case, can you elaborate what cases of update/delete-many do you have? What type of predicates do you use?

In my application, there are some types that operations like update/delete-many are not allowed, and we accept only entity-specific mutations like DeleteOne, UpdateOne.

But in the cases that I do want to execute mutations only for entities in the same tenant - instead of querying the database (for getting the tenant), I'm just adding another generic predicate for filtering by the tenant-edge (using the [entql](https://entgo.io/docs/feature-flags/#entql-filtering) feature). There's some [example](https://github.com/facebook/ent/blob/master/examples/privacytenant/rule/rule.go#L58) in the `examples` directory for read operations. 

I'll try to add code examples here later today.

**Comment by a8m** [comment]
I've added to the website another [example](https://entgo.io/docs/privacy/#multi-tenancy) based on this issue (see it at the ~end of the page). Please update to latest if you want to run and use this example.

Thanks and let me know if you still need help with it.
